common/dal2/src/wraps/seekable.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::future::Future;
+use std::io;
+use std::io::SeekFrom;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+use futures::pin_mut;
+use futures::ready;
+use futures::AsyncRead;
+use futures::AsyncSeek;
+use futures::FutureExt;
+
+use crate::error::Result;
+use crate::Reader;
+
+/// If we already know a file's total size, we can implement Seek for it.
+///
+/// - Every time we call `read` we will send a new http request to fetch data from cloud storage like s3.
+/// - Every time we call `seek` we will update the `pos` field just in memory.
+/// # NOTE
+///
+/// It's better to use SeekableReader as an inner reader inside BufReader.
+///
+/// # TODO
+///
+/// We need use update the metrics.
+pub struct SeekableReader {
+    op: crate::Operator,
+    key: String,
+    total: u64,
```

Comment:
As you may have noticed,  the field `total: u64` somehow implies that the reader should know the size of the file/object being read in advance (or during the construction of `SeekableReader`). This might not be necessary in some cases, e.g. just reading a file till EOF.
Just a reminder, we can refine this in subsequent PRs if necessary.

Reply:
Oh, that makes sense. I tracked this issue at https://github.com/datafuselabs/databend/issues/3677